### PR TITLE
Some new minor improvements

### PR DESCRIPTION
**Pull request by IMAN4K**
Signed-off-by: Iman Ahmadvand iman72411@gmail.com (github: @IMAN4K)

### Changes
* Adding support for both `SoftwareSerial` & `HardwareSerial`
* New logging method with `[A6lib]` stamp
* Some typo changes (e.g `SMSmessage` -> `SMSInfo`)
* Start documenting source code via `Doxygen`

### CODE REVIEW

**Comment by skorokithakis** [comment]
Please only make one change per PR, this is making reviewing hard by not keeping things tidy. Also, you have some git output included in the file, so this won't compile.

**Comment by IMAN4K** [comment]
I remove the git stamps from source, now it will compile as fine.
Sorry but i can not divide the commits themselves :(
If you have any problem review changes, or these commits make `A6lib` broken just let me know to close this PR and the next ones!

**Comment by skorokithakis** [comment]
Ah, that's fine, I will try to review. The problem is that I don't have a testing board with me so I can't actually check if it works or not, I can only read the code and try to figure it out. I'll try reviewing now, thank you!

File: A6lib.cpp [hunk 1]
```diff
@@ -1,319 +1,368 @@
...
+#define A6_CMD_TIMEOUT 2000
 
-A6lib::A6lib(int transmitPin, int receivePin) {
-#ifdef ESP8266
-    A6conn = new SoftwareSerial(receivePin, transmitPin, false, 1024);
-#else
-    A6conn = new SoftwareSerial(receivePin, transmitPin, false);
+#ifdef DEBUG
+#ifndef DEBUG_PORT
+#	define DEBUG_PORT Serial
```
Comment (skorokithakis) [inline A6lib.cpp]: Should there be a tab between `#` and `define`?

File: A6lib.cpp [hunk 2]
```diff
@@ -1,319 +1,368 @@
-#include <Arduino.h>
```
Comment (skorokithakis) [inline A6lib.cpp]: Does removing these imports influence anything? What board are you running this on?

**Comment by IMAN4K** [comment]
That's why i said to create new `dev` branch ;)
That way users can test & verify the `dev` version of library around some time and make it even more stable.

**Comment by skorokithakis** [comment]
Yeah, unfortunately there aren't many devs that can try it out :/ Still, it sounds like a good idea for this PR. Let me create one and we can re-select that as the target here.

File: A6lib.cpp (hunk 2 above)
Comment (IMAN4K) [reply A6lib.cpp]: Not actually.
I just removed it in other upcoming commits
I did all test with `Nodemcu` on my desk

File: A6lib.cpp (hunk 1 above)
Comment (IMAN4K) [reply A6lib.cpp]: Just a habit :)
Will remove it

**Comment by IMAN4K** [comment]
Things get complicated because of `git cherry-pick`! and last changes in master. i modified last commit and this(`sms` example) should be compiled just fine:
```
#include "A6lib.h"

#ifdef ESP8266
#define D0 0
#define D5 8
#define D6 7
#endif

// Instantiate the library with TxPin, RxPin.
A6lib A6l(D6, D5);

int unreadSMSLocs[30] = {0};
int unreadSMSNum = 0;
SMSInfo sms;

void setup() {
    Serial.begin(115200);

    delay(1000);

    // Power-cycle the module to reset it.
    A6l.powerCycle(D0);
    A6l.blockUntilReady(9600);
}

void loop() {
    String myNumber = "+1132352890";

    callInfo cinfo = A6l.checkCallStatus();
    if (cinfo.direction == DIR_INCOMING) {
        if (myNumber.endsWith(cinfo.number)) {
            // If the number that sent the SMS is ours, reply.
            A6l.sendSMS(myNumber, "I can't come to the phone right now, I'm a machine.");
            A6l.hangUp();
        }

        // Get the memory locations of unread SMS messages.
        unreadSMSNum = A6l.getUnreadSMSLocs(unreadSMSLocs, 30);

        for (int i = 0; i < unreadSMSNum; i++) {
            Serial.print("New message at index: ");
            Serial.println(unreadSMSLocs[i], DEC);

            sms = A6l.readSMS(unreadSMSLocs[i]);
            Serial.println(sms.number);
            Serial.println(sms.date);
            Serial.println(sms.message);
        }
        delay(1000);
    }
}
```

**Comment by skorokithakis** [comment]
That compiled, thanks!

File: A6lib.cpp [hunk 3]
```diff
@@ -368,99 +417,113 @@ void A6lib::enableSpeaker(byte enable) {
...
+		rate = rates[i];
+
+		Serial1.printf("sumping ports");
+		if (ports.hport != nullptr)
+			Serial.println("hardware ports is present....");
+		if (ports.sport != nullptr)
+			Serial1.println("software port is present...");
+
+		if (ports.hport != nullptr)
+			ports.hport->begin(rate);
```
Comment (skorokithakis) [inline A6lib.cpp]: Couldn't this go in the `if` checks above?

File: A6lib.h [hunk 4]
```diff
@@ -1,76 +1,69 @@
...
+	MODE_VOICE_AND_FAX_FMODE = 8,
+	MODE_UNKNOWN = 9
 };
 
-struct SMSmessage {
-    String number;
-    String date;
-    String message;
+struct SMSInfo {
+	SMSInfo() : number{}, date{}, message{} {
```
Comment (skorokithakis) [inline A6lib.h]: Why the renaming of this struct?

File: A6lib.cpp (hunk 3 above)
Comment (IMAN4K) [reply A6lib.cpp]: I Fix all these in next commits and it's perfect:)
I recommend to delay this review until next one.

**Comment by IMAN4K** [comment]
And BTW after all my commits get pushed we need to bring `Travis-CI` auto build to this repo

File: A6lib.h (hunk 4 above)
Comment (skorokithakis) [reply A6lib.h]: Ahh, I see, that makes sense. Thanks.

**Comment by skorokithakis** [comment]
@IMAN4K Yes, do you have any files that do this already? I would love to get Travis building this.

**Comment by IMAN4K** [comment]
This one is straight-forward
* [`.travis.yml`](https://github.com/me-no-dev/ESPAsyncTCP/blob/master/.travis.yml)
* [`common.sh`](https://github.com/me-no-dev/ESPAsyncTCP/blob/master/travis/common.sh)

**Comment by skorokithakis** [comment]
Oh wow, that's pretty big :stuck_out_tongue: I was actually in the middle of doing this with PlatformIO which doesn't need a screen and is much faster, I just need to specify the paths. I think I'll be done in a few minutes, thanks!

**Comment by skorokithakis** [comment]
Alright, added! If you merge `master` in, it should start working!

**Comment by IMAN4K** [comment]
Great
Could you please duplicate changes in `develop`?

**Comment by skorokithakis** [comment]
I've taken the liberty of reformatting and adding the Travis file in this branch, it should build now!

**Comment by skorokithakis** [comment]
I think it's good to merge, as soon as you fix the `Serial1` statement and the `if` check we're good to go!

**Comment by IMAN4K** [comment]
I'm on it!

**Comment by skorokithakis** [comment]
Looks like CI failed because the SMS example is wrong, it still uses `SMSMessage`. If you could fix the example as well, I'd be grateful.

**Comment by IMAN4K** [comment]
@skorokithakis Any news !?

**Comment by IMAN4K** [comment]
Formatting Done!

**Comment by skorokithakis** [comment]
Merged, thank you very much for your help!

**Pull request merged**